metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala

```diff
@@ -344,7 +345,27 @@ object MetalsEnrichments
 
     def writeText(text: String): Unit = {
       path.parent.createDirectories()
-      Files.write(path.toNIO, text.getBytes(StandardCharsets.UTF_8))
+      val tmp = Files.createTempFile("metals", path.filename)
+      // Write contents first to a temporary file and then try to
+      // atomically move the file to the destination. The atomic move
+      // reduces the risk that another tool will concurrently read the
+      // file contents during a half-complete file write.
+      Files.write(
+        tmp,
+        text.getBytes(StandardCharsets.UTF_8),
+        StandardOpenOption.TRUNCATE_EXISTING
+      )
+      try {
+        Files.move(
+          tmp,
+          path.toNIO,
+          StandardCopyOption.REPLACE_EXISTING,
+          StandardCopyOption.ATOMIC_MOVE
+        )
+      } catch {
+        case NonFatal(_) =>
+          Files.move(tmp, path.toNIO, StandardCopyOption.REPLACE_EXISTING)
```

Comment:
So there is still a risk we will have a non-atomic operation? Shouldn't we retry the atomic move again or resign altogether.

Reply:
Some file systems don’t support atomic moves so the best we can do is fall back to a normal move.

Reply:
Could there be a more specific exception to catch here?

Reply:
The docs don't specify what exception will be thrown so I didn't want to be too specific.